File: full-stack-lab/sdk/java/src/main/java/org/zo/interop/EchoServer.java
package org.zo.interop;

import org.openziti.ZitiAddress;
import org.openziti.ZitiContext;

import java.nio.ByteBuffer;
import java.nio.channels.AsynchronousServerSocketChannel;
import java.nio.channels.AsynchronousSocketChannel;

/**
 * Echo server: binds ZITI_SERVICE on the overlay and echoes bytes back forever.
 * No TCP listener, no host port.
 */
final class EchoServer {

    static void run() throws Exception {
        ZitiContext ctx = Env.context();
        String service = Env.service();

        AsynchronousServerSocketChannel server = ctx.openServer();
        server.bind(new ZitiAddress.Bind(service));
        System.err.println("echo server bound service '" + service + "' (lang=" + Env.LANG + ")");

        while (true) {
            AsynchronousSocketChannel conn = server.accept().get();
            Thread t = new Thread(() -> handle(conn));
            t.setDaemon(true);
            t.start();
        }
    }

    private static void handle(AsynchronousSocketChannel conn) {
        try (conn) {
            ByteBuffer buf = ByteBuffer.allocate(4096);
            while (true) {
                buf.clear();
                int n = conn.read(buf).get();
                if (n < 0) {
                    break;
                }
                buf.flip();
                while (buf.hasRemaining()) {
                    conn.write(buf).get();
                }
            }
        } catch (Exception e) {
            System.err.println("echo conn error: " + e.getMessage());
        }
    }
}

File: full-stack-lab/sdk/csharp/EchoServer.cs
using System.IO;
using System.Net.Sockets;
using OpenZiti;

namespace ZoSdk;

/// <summary>
/// Binds ZITI_SERVICE and echoes every byte received back to the sender, forever.
/// </summary>
public static class EchoServer
{
    public static void Run()
    {
        string identityFile = Env.IdentityFile;
        string serviceName  = Env.ServiceName;

        Console.Error.WriteLine($"[echo-server] identity={identityFile} service={serviceName}");

        var ctx    = Env.LoadContext(identityFile);
        var server = new ZitiSocket(SocketType.Stream);

        API.Bind(server, ctx, serviceName, "");
        API.Listen(server, 32);

        Console.Error.WriteLine($"[echo-server] listening on service '{serviceName}'");

        while (true)
        {
            ZitiSocket client = API.Accept(server, out string caller);
            Console.Error.WriteLine($"[echo-server] accepted from {caller}");
            // Handle each connection on a thread-pool thread so the accept loop keeps going.
            _ = Task.Run(() => HandleClient(client, caller));
        }
    }

    private static void HandleClient(ZitiSocket socket, string caller)
    {
        try
        {
            using var ns  = socket.ToNetworkStream();
            var buf       = new byte[4096];
            int n;
            while ((n = ns.Read(buf, 0, buf.Length)) > 0)
            {
                ns.Write(buf, 0, n);
            }
            Console.Error.WriteLine($"[echo-server] {caller} disconnected");
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine($"[echo-server] error from {caller}: {ex.Message}");
        }
        finally
        {
            socket.Dispose();
        }
    }
}
